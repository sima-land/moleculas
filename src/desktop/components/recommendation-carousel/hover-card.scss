.card {
  position: absolute;
  opacity: 1;
  visibility: visible;
  &.hidden {
    opacity: 0;
    visibility: hidden;
    pointer-events: none;
  }
}
