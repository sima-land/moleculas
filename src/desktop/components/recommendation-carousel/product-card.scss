@use 'node_modules/@dev-dep/ui-nucleons/colors';

.root {
  padding: 16px 16px 24px;
  background: #fff;
}

.control-wrapper {
  display: flex;
  flex-wrap: wrap;
  margin-top: 16px;
}

.unit-text {
  padding-top: 12px; // чтобы при переносе на следующую строку правильно отступил от кнопки
  font-size: 12px;
  line-height: 16px;
  color: colors.$gray38;
}

.bottom-text {
  margin-top: 12px;
  font-size: 12px;
  line-height: 16px;
  color: colors.$gray38;
}
