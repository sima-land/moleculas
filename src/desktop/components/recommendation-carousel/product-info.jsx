import React from 'react';
import { Link } from '@dev-dep/ui-nucleons/link';
import { Price } from '@dev-dep/ui-nucleons/price';
import { Image } from '../../../common/components/image';
import { isNonEmptyArray } from '../../../common/helpers/is-non-empty';
import { Badge } from '../../../common/components/badge';
import classnames from 'classnames/bind';
import styles from './product-info.scss';
import { WishButton } from '../../../common/components/wish-button';
import { QuickViewButton } from '../../../common/components/quick-view-button';
import withPrevent from '@dev-dep/ui-nucleons/helpers/with-prevent';

const cx = classnames.bind(styles);

/**
 * Компонент базовой информации о товаре.
 * @param {Object} props Свойства компонента.
 * @return {ReactElement} Элемент.
 */
export const ProductInfo = ({
  badges,
  currencyGrapheme,
  favorite,
  imageSrc,
  name,
  oldPrice,
  onFavoriteClick,
  onQuickViewClick,
  price,
  url,
  withImageButtons,
}) => (
  <>
    <a className={cx('row', 'image-row')} href={url}>
      <Image
        className={cx('image')}
        alt={name}
        src={imageSrc}
      />
      {withImageButtons && (
        <>
          <WishButton
            checked={favorite}
            className={cx('on-image-btn', 'wish')}
            onClick={withPrevent(onFavoriteClick)}
          />
          <QuickViewButton
            className={cx('on-image-btn', 'quick-view')}
            onClick={withPrevent(onQuickViewClick)}
          />
        </>
      )}
    </a>

    {Boolean(price) && (
      <div className={cx('row', 'prices-row')}>
        <Price
          value={price}
          currencyGrapheme={currencyGrapheme}
          className={cx('price')}
        />
        {Boolean(oldPrice) && (
          <Price
            value={oldPrice}
            currencyGrapheme={currencyGrapheme}
            className={cx('old-price')}
            crossedOut
          />
        )}
      </div>
    )}

    {isNonEmptyArray(badges) && (
      <div className={cx('row', 'badges-row')}>
        {badges.map((badge, index) => (
          <Badge
            {...badge}
            key={index}
            className={cx('badge')}
          />
        ))}
      </div>
    )}

    <Link
      className={cx('row', 'link')}
      href={url}
      children={name}
      color='gray87'
    />
  </>
);
