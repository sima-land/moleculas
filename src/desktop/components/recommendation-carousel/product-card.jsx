import React, { forwardRef } from 'react';
import { BoxShadow as Shadow } from '@dev-dep/ui-nucleons/styling/shadows';
import { MediumRounds as Rounds } from '@dev-dep/ui-nucleons/styling/shapes';
import { ProductInfo } from './product-info';
import { InCartControl } from '../../../common/components/in-cart-control';
import classnames from 'classnames/bind';
import styles from './product-card.scss';

const cx = classnames.bind(styles);

/**
 * Карточка рекомендованного товара.
 * @param {Object} props Свойства.
 * @param {import('./types').ProductData} props.productInfo Данные товара.
 */
export const ProductCard = forwardRef(({
  className,
  inCartControl,
  productInfo,
  ...divProps
}, ref) => (
  <div
    {...divProps}
    ref={ref}
    className={cx('root', Rounds.all, Shadow.z4, className)}
  >
    <ProductInfo
      {...productInfo}
      withImageButtons
    />

    <div className={cx('control-wrapper')}>
      <InCartControl {...inCartControl} className={cx('control')} />

      {inCartControl && inCartControl.stepText && (
        <div className={cx('unit-text')}>
          {inCartControl.stepText}
        </div>
      )}

      {inCartControl && inCartControl.markupText && (
        <div className={cx('bottom-text')}>
          {inCartControl.markupText}
        </div>
      )}
    </div>
  </div>
));
