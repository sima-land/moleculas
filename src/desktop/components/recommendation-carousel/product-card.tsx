import React, { forwardRef } from 'react';
import { BoxShadow as Shadow } from '@sima-land/ui-nucleons/styling/shadows';
import { MediumRounds as Rounds } from '@sima-land/ui-nucleons/styling/shapes';
import { ProductInfo, ProductInfoProps } from './product-info';
import { InCartControl, InCartControlProps } from '../../../common/components/in-cart-control';
import classnames from 'classnames/bind';
import styles from './product-card.module.scss';
import { CartItem } from './types';

export interface ProductCardProps extends React.HTMLAttributes<HTMLDivElement> {
  inCartControl: CartItem & InCartControlProps
  productInfo: Omit<ProductInfoProps, 'withQuickView'>
}

const cx = classnames.bind(styles);

/**
 * Карточка рекомендованного товара.
 */
export const ProductCard = forwardRef<HTMLDivElement | null, ProductCardProps>(({
  className,
  inCartControl,
  productInfo,
  ...divProps
}, ref) => (
  <div
    {...divProps}
    ref={ref}
    className={cx('root', Rounds.all, Shadow.z4, className)}
  >
    <ProductInfo
      {...productInfo}
      withQuickView
    />

    <div className={cx('control-wrapper')}>
      <InCartControl {...inCartControl} className={cx('control')} />

      {inCartControl && inCartControl.stepText && (
        <div className={cx('unit-text')}>
          {inCartControl.stepText}
        </div>
      )}

      {inCartControl && inCartControl.markupText && (
        <div className={cx('bottom-text')}>
          {inCartControl.markupText}
        </div>
      )}
    </div>
  </div>
));
