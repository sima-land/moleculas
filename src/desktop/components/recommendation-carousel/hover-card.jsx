import React, { useRef, useState } from 'react';
import { getOriginCorrection } from '@dev-dep/ui-nucleons/with-tooltip/utils';
import { ProductCard } from './product-card';
import { omit, pick } from 'lodash';
import classnames from 'classnames/bind';
import styles from './hover-card.scss';

const cx = classnames.bind(styles);

/**
 * Карточка товара, появляющаяся над элементом карусели.
 * @param {Object} props Свойства компонента.
 * @return {ReactElement} Элемент.
 */
export const HoverCard = ({
  controlRef,
  onAdd,
  onChange,
  onFavoriteClick,
  onQuickViewClick,
  onSubtract,
  ...restProps
}) => {
  const ref = useRef();
  const [itemInfo, setItemInfo] = useState();

  const Control = {
    show: (info, target) => {
      setItemInfo(info);

      const element = ref.current;
      const rect = target.getBoundingClientRect();
      const correction = getOriginCorrection(ref.current);

      element.classList.remove(cx('hidden'));
      element.style.left = `${correction.x + rect.left - 16}px`;
      element.style.top = `${correction.y + rect.top - 16}px`;
      element.style.width = `${rect.width + 32}px`;
    },

    hide: () => {
      // не нужно сбрасывать данные тут (например setInfo(null)) так как blur не будет работать из-за размонтирования
      ref.current && ref.current.classList.add(cx('hidden'));
    },
  };

  // eslint-disable-next-line require-jsdoc
  const bindInfo = fn => () => fn && fn(itemInfo);

  // отдаем управление чтобы не поднимать состояние выше
  controlRef.current = Control;

  const [cartData, productData] = [pick, omit].map(fn => fn(itemInfo, [
    'allowFloat',
    'canAdd',
    'canSubtract',
    'inCart',
    'isFetching',
    'markupText',
    'qty',
    'stepText',
  ]));

  return (
    <ProductCard
      {...restProps}
      ref={ref}
      className={cx('card', 'hidden')}
      onMouseLeave={Control.hide}
      inCartControl={{
        ...cartData,
        onAdd: bindInfo(onAdd),
        onChange: newQty => onChange && onChange(itemInfo, newQty),
        onSubtract: bindInfo(onSubtract),
      }}
      productInfo={{
        ...productData,
        onFavoriteClick: bindInfo(onFavoriteClick),
        onQuickViewClick: bindInfo(onQuickViewClick),
      }}
    />
  );
};
