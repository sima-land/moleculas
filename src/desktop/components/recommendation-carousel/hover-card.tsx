import React, { useLayoutEffect, useRef, useState } from 'react';
import { getOriginCorrection } from '@sima-land/ui-nucleons/with-tooltip/utils';
import { ProductCard, ProductCardProps } from './product-card';
import classnames from 'classnames/bind';
import styles from './hover-card.module.scss';
import { CarouselItem } from './types';

export interface HoverCardProps extends Omit<ProductCardProps, 'onChange' | 'inCartControl' | 'productInfo'> {
  info: CarouselItem
  targetRef: React.RefObject<HTMLElement | null>
  onAdd?: (info: CarouselItem) => void
  onChange?: (info: CarouselItem, newQty: number) => void
  onMouseLeave?: () => void
  onQuickViewClick?: (info: CarouselItem) => void
  onSubtract?: (info: CarouselItem) => void
}

const cx = classnames.bind(styles);

/**
 * Карточка товара, появляющаяся над элементом карусели.
 * @param props Свойства компонента.
 * @return Элемент.
 */
export const HoverCard = ({
  info,
  targetRef,
  onAdd,
  onChange,
  onMouseLeave,
  onQuickViewClick,
  onSubtract,
  ...restProps
}: HoverCardProps) => {
  const ref = useRef<HTMLDivElement | null>(null);
  const [ready, setReady] = useState<boolean>(false);

  useLayoutEffect(() => {
    if (info && targetRef.current && ref.current) {
      const element = ref.current;
      const rect = targetRef.current.getBoundingClientRect();
      const correction = getOriginCorrection(ref.current);

      element.style.left = `${correction.x + rect.left - 16}px`;
      element.style.top = `${correction.y + rect.top - 16}px`;
      element.style.width = `${rect.width + 32}px`;

      setReady(true);
    }
  }, [info]);

  // eslint-disable-next-line require-jsdoc
  const bindInfo = (fn: any) => () => fn && fn(info);

  return (
    <ProductCard
      {...restProps}
      ref={ref}
      className={cx('card', !ready && 'hidden')}
      onMouseLeave={() => {
        setReady(false);
        onMouseLeave && onMouseLeave();
      }}
      inCartControl={{
        ...info.cart,
        onAdd: bindInfo(onAdd),
        onChange: (newQty: any) => onChange && onChange(info, newQty),
        onSubtract: bindInfo(onSubtract),
      } as any}
      productInfo={{
        ...info.product,
        onQuickViewClick: bindInfo(onQuickViewClick),
      } as any}
    />
  );
};
