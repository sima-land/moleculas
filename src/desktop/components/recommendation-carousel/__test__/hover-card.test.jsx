import React, { createRef } from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { HoverCard } from '../hover-card';
import { ProductInfo } from '../product-info';
import { Link } from '@dev-dep/ui-nucleons/link';
import { ProductCard } from '../product-card';
import { Stepper } from '@dev-dep/ui-nucleons/stepper';

describe('<HoverCard />', () => {
  it('should renders correctly', () => {
    const wrapper = mount(
      <HoverCard
        controlRef={createRef()}
        onQuickViewClick={jest.fn()}
        onFavoriteClick={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle image button click callbacks', () => {
    const Spy = {
      onQuickViewClick: jest.fn(),
      onFavoriteClick: jest.fn(),
    };

    const wrapper = mount(
      <HoverCard
        controlRef={createRef()}
        {...Spy}
      />
    );

    // quick view
    expect(Spy.onQuickViewClick).toBeCalledTimes(0);
    act(() => {
      wrapper.find('[data-testid="quick-view-button"]').simulate('click');
    });
    expect(Spy.onQuickViewClick).toBeCalledTimes(1);

    // favorite
    expect(Spy.onFavoriteClick).toBeCalledTimes(0);
    act(() => {
      wrapper.find('[data-testid="favorite-button"]').simulate('click');
    });
    expect(Spy.onFavoriteClick).toBeCalledTimes(1);
  });

  it('should set control to ref', () => {
    const controlRef = createRef();

    const itemInfo = {
      url: 'https://www.sima-land.ru/123456',
      name: 'Ножницы портновские, 10, 26 см, цвет чёрный',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476',
      price: 225,
      oldPrice: 250,
      favorite: false,
      currencyGrapheme: '₽',
      withImageButtons: true,
      onQuickViewClick: jest.fn(),
      onFavoriteClick: jest.fn(),
      markupText: 'Комплектация + 50 ₽ при покупке до 20 шт',
    };

    const wrapper = mount(
      <HoverCard
        controlRef={controlRef}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ProductCard).getDOMNode().classList.contains('hidden')).toBe(true);

    act(() => {
      const fakeTarget = document.createElement('div');
      controlRef.current.show(itemInfo, fakeTarget);
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ProductInfo).find(Link).text()).toBe(itemInfo.name);
    expect(wrapper.find(ProductCard).getDOMNode().classList.contains('hidden')).toBe(false);

    act(() => {
      wrapper.find(ProductCard).simulate('mouseleave');
    });
    wrapper.update();

    expect(wrapper.find(ProductCard).getDOMNode().classList.contains('hidden')).toBe(true);
  });

  it('should handle cart data/control props', () => {
    const controlRef = createRef();

    const Spy = {
      onAdd: jest.fn(),
      onChange: jest.fn(),
      onSubtract: jest.fn(),
    };

    const productInfo = {
      name: 'Test Product',
      imageSrc: 'http://image.com/test',
      imageAlt: 'Alternative text for test image',
      url: 'https://www.sima-land.ru/123',
      price: 100,
      currencyGrapheme: '$',
      oldPrice: 200,
      stepText: 'по 1 шт',
      markupText: 'Комплектация + 50$ при покупке до 20 шт',
      qty: 5,
    };

    const wrapper = mount(
      <HoverCard
        controlRef={controlRef}
        onAdd={Spy.onAdd}
        onChange={Spy.onChange}
        onSubtract={Spy.onSubtract}
      />
    );

    act(() => {
      const fakeTarget = document.createElement('div');
      controlRef.current.show(productInfo, fakeTarget);
    });
    wrapper.update();

    expect(wrapper.find(HoverCard).find(Stepper).find('[data-testid="stepper:input"]').prop('value')).toBe(5);

    Object.values(Spy).forEach(fn => {
      expect(fn).toBeCalledTimes(0);
    });

    // add
    act(() => {
      wrapper.find(HoverCard).find(Stepper).find('svg[data-testid="stepper:plus"]').simulate('click');
    });
    wrapper.update();

    expect(Spy.onAdd).toBeCalledTimes(1);

    // subtract
    act(() => {
      wrapper.find(HoverCard).find(Stepper).find('svg[data-testid="stepper:minus"]').simulate('click');
    });
    wrapper.update();

    expect(Spy.onSubtract).toBeCalledTimes(1);

    // subtract
    act(() => {
      wrapper.find(HoverCard).find(Stepper).find('[data-testid="stepper:input"]').simulate('blur', {
        target: {
          value: 15,
        },
      });
    });
    wrapper.update();

    expect(Spy.onChange).toBeCalledTimes(1);
  });
});
