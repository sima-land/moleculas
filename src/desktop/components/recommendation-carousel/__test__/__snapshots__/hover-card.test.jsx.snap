// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HoverCard /> should renders correctly 1`] = `
<HoverCard
  controlRef={
    Object {
      "current": Object {
        "hide": [Function],
        "show": [Function],
      },
    }
  }
  onFavoriteClick={[MockFunction]}
  onQuickViewClick={[MockFunction]}
>
  <ForwardRef
    className="card hidden"
    inCartControl={
      Object {
        "onAdd": [Function],
        "onChange": [Function],
        "onSubtract": [Function],
      }
    }
    onMouseLeave={[Function]}
    productInfo={
      Object {
        "onFavoriteClick": [Function],
        "onQuickViewClick": [Function],
      }
    }
  >
    <div
      className="root rounds-md__all shadow-z4 card hidden"
      onMouseLeave={[Function]}
    >
      <ProductInfo
        onFavoriteClick={[Function]}
        onQuickViewClick={[Function]}
        withImageButtons={true}
      >
        <a
          className="row image-row"
        >
          <Image
            className="image"
          >
            <div
              className="root image"
            >
              <img />
            </div>
          </Image>
          <WithHint
            text="Добавить в избранное"
          >
            <ForwardRef
              className="on-image-btn wish"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <ForwardRef(SvgHeart)
                className="button on-image-btn wish"
                data-testid="favorite-button"
                height={24}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="button"
                title="Добавить в избранное"
                width={24}
              >
                <svg
                  className="button on-image-btn wish"
                  data-testid="favorite-button"
                  height={24}
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="button"
                  title="Добавить в избранное"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.944 3.6h-.001c-2.7-1.879-5.978-1.165-7.943.793-1.964-1.958-5.241-2.683-7.943-.792C2.524 4.673 1.57 6.473 1.504 8.369c-.075 2.145.826 4.042 2.36 5.985 1.512 1.912 3.685 3.927 6.261 6.315l.086.08h.002l.098.091a2.466 2.466 0 003.371-.012l.107-.1.027-.025c2.602-2.407 4.795-4.436 6.316-6.361 1.534-1.941 2.435-3.838 2.364-5.982v-.001c-.065-1.886-1.02-3.686-2.552-4.758zM12 18.274c-2.278-2.105-4.14-3.84-5.45-5.401-1.358-1.62-2.047-2.968-2.047-4.261 0-1.774 1.285-3.07 2.999-3.07 1.055 0 2.103.553 2.72 1.39.38.516.996 1.017 1.78 1.017.789 0 1.406-.507 1.787-1.025.612-.831 1.657-1.382 2.71-1.382 1.713 0 2.998 1.296 2.998 3.07 0 1.293-.689 2.641-2.047 4.26-1.309 1.561-3.171 3.296-5.45 5.402z"
                  />
                </svg>
              </ForwardRef(SvgHeart)>
            </ForwardRef>
          </WithHint>
          <WithHint
            text="Быстрый просмотр"
          >
            <ForwardRef
              className="on-image-btn quick-view"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <ForwardRef(SvgQuickView)
                className="root on-image-btn quick-view"
                data-testid="quick-view-button"
                height={24}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="button"
                title="Быстрый просмотр"
                width={24}
              >
                <svg
                  className="root on-image-btn quick-view"
                  data-testid="quick-view-button"
                  height={24}
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="button"
                  title="Быстрый просмотр"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.94 22.06l-3.219-3.217-.339.22A8.956 8.956 0 0111.5 20.5a9 9 0 119-9c0 1.8-.528 3.476-1.438 4.882l-.22.34 3.219 3.217a1.5 1.5 0 01-2.122 2.122zM9 10h1V9a1.5 1.5 0 113 0v1h1a1.5 1.5 0 110 3h-1v1a1.5 1.5 0 11-3 0v-1H9a1.5 1.5 0 110-3zm2.5-4.5a6 6 0 100 12 6 6 0 000-12z"
                  />
                </svg>
              </ForwardRef(SvgQuickView)>
            </ForwardRef>
          </WithHint>
        </a>
        <ForwardRef(Link)
          className="row link"
          color="gray87"
        >
          <a
            className="link row link color__gray87 hover-color__gray54"
          />
        </ForwardRef(Link)>
      </ProductInfo>
      <div
        className="control-wrapper"
      >
        <InCartControl
          className="control"
          onAdd={[Function]}
          onChange={[Function]}
          onSubtract={[Function]}
        >
          <ForwardRef(Button)
            className="control control"
            onClick={[Function]}
            size="small"
          >
            <button
              className="control control button-base button-primary button-small text-button-small"
              onClick={[Function]}
            >
              В корзину
            </button>
          </ForwardRef(Button)>
        </InCartControl>
      </div>
    </div>
  </ForwardRef>
</HoverCard>
`;

exports[`<HoverCard /> should set control to ref 1`] = `
<HoverCard
  controlRef={
    Object {
      "current": Object {
        "hide": [Function],
        "show": [Function],
      },
    }
  }
>
  <ForwardRef
    className="card hidden"
    inCartControl={
      Object {
        "onAdd": [Function],
        "onChange": [Function],
        "onSubtract": [Function],
      }
    }
    onMouseLeave={[Function]}
    productInfo={
      Object {
        "onFavoriteClick": [Function],
        "onQuickViewClick": [Function],
      }
    }
  >
    <div
      className="root rounds-md__all shadow-z4 card hidden"
      onMouseLeave={[Function]}
    >
      <ProductInfo
        onFavoriteClick={[Function]}
        onQuickViewClick={[Function]}
        withImageButtons={true}
      >
        <a
          className="row image-row"
        >
          <Image
            className="image"
          >
            <div
              className="root image"
            >
              <img />
            </div>
          </Image>
          <WithHint
            text="Добавить в избранное"
          >
            <ForwardRef
              className="on-image-btn wish"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <ForwardRef(SvgHeart)
                className="button on-image-btn wish"
                data-testid="favorite-button"
                height={24}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="button"
                title="Добавить в избранное"
                width={24}
              >
                <svg
                  className="button on-image-btn wish"
                  data-testid="favorite-button"
                  height={24}
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="button"
                  title="Добавить в избранное"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.944 3.6h-.001c-2.7-1.879-5.978-1.165-7.943.793-1.964-1.958-5.241-2.683-7.943-.792C2.524 4.673 1.57 6.473 1.504 8.369c-.075 2.145.826 4.042 2.36 5.985 1.512 1.912 3.685 3.927 6.261 6.315l.086.08h.002l.098.091a2.466 2.466 0 003.371-.012l.107-.1.027-.025c2.602-2.407 4.795-4.436 6.316-6.361 1.534-1.941 2.435-3.838 2.364-5.982v-.001c-.065-1.886-1.02-3.686-2.552-4.758zM12 18.274c-2.278-2.105-4.14-3.84-5.45-5.401-1.358-1.62-2.047-2.968-2.047-4.261 0-1.774 1.285-3.07 2.999-3.07 1.055 0 2.103.553 2.72 1.39.38.516.996 1.017 1.78 1.017.789 0 1.406-.507 1.787-1.025.612-.831 1.657-1.382 2.71-1.382 1.713 0 2.998 1.296 2.998 3.07 0 1.293-.689 2.641-2.047 4.26-1.309 1.561-3.171 3.296-5.45 5.402z"
                  />
                </svg>
              </ForwardRef(SvgHeart)>
            </ForwardRef>
          </WithHint>
          <WithHint
            text="Быстрый просмотр"
          >
            <ForwardRef
              className="on-image-btn quick-view"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <ForwardRef(SvgQuickView)
                className="root on-image-btn quick-view"
                data-testid="quick-view-button"
                height={24}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="button"
                title="Быстрый просмотр"
                width={24}
              >
                <svg
                  className="root on-image-btn quick-view"
                  data-testid="quick-view-button"
                  height={24}
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="button"
                  title="Быстрый просмотр"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.94 22.06l-3.219-3.217-.339.22A8.956 8.956 0 0111.5 20.5a9 9 0 119-9c0 1.8-.528 3.476-1.438 4.882l-.22.34 3.219 3.217a1.5 1.5 0 01-2.122 2.122zM9 10h1V9a1.5 1.5 0 113 0v1h1a1.5 1.5 0 110 3h-1v1a1.5 1.5 0 11-3 0v-1H9a1.5 1.5 0 110-3zm2.5-4.5a6 6 0 100 12 6 6 0 000-12z"
                  />
                </svg>
              </ForwardRef(SvgQuickView)>
            </ForwardRef>
          </WithHint>
        </a>
        <ForwardRef(Link)
          className="row link"
          color="gray87"
        >
          <a
            className="link row link color__gray87 hover-color__gray54"
          />
        </ForwardRef(Link)>
      </ProductInfo>
      <div
        className="control-wrapper"
      >
        <InCartControl
          className="control"
          onAdd={[Function]}
          onChange={[Function]}
          onSubtract={[Function]}
        >
          <ForwardRef(Button)
            className="control control"
            onClick={[Function]}
            size="small"
          >
            <button
              className="control control button-base button-primary button-small text-button-small"
              onClick={[Function]}
            >
              В корзину
            </button>
          </ForwardRef(Button)>
        </InCartControl>
      </div>
    </div>
  </ForwardRef>
</HoverCard>
`;

exports[`<HoverCard /> should set control to ref 2`] = `
<HoverCard
  controlRef={
    Object {
      "current": Object {
        "hide": [Function],
        "show": [Function],
      },
    }
  }
>
  <ForwardRef
    className="card hidden"
    inCartControl={
      Object {
        "markupText": "Комплектация + 50 ₽ при покупке до 20 шт",
        "onAdd": [Function],
        "onChange": [Function],
        "onSubtract": [Function],
      }
    }
    onMouseLeave={[Function]}
    productInfo={
      Object {
        "currencyGrapheme": "₽",
        "favorite": false,
        "imageSrc": "https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476",
        "name": "Ножницы портновские, 10, 26 см, цвет чёрный",
        "oldPrice": 250,
        "onFavoriteClick": [Function],
        "onQuickViewClick": [Function],
        "price": 225,
        "url": "https://www.sima-land.ru/123456",
        "withImageButtons": true,
      }
    }
  >
    <div
      className="root rounds-md__all shadow-z4 card hidden"
      onMouseLeave={[Function]}
    >
      <ProductInfo
        currencyGrapheme="₽"
        favorite={false}
        imageSrc="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476"
        name="Ножницы портновские, 10, 26 см, цвет чёрный"
        oldPrice={250}
        onFavoriteClick={[Function]}
        onQuickViewClick={[Function]}
        price={225}
        url="https://www.sima-land.ru/123456"
        withImageButtons={true}
      >
        <a
          className="row image-row"
          href="https://www.sima-land.ru/123456"
        >
          <Image
            alt="Ножницы портновские, 10, 26 см, цвет чёрный"
            className="image"
            src="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476"
          >
            <div
              className="root image"
            >
              <img
                alt="Ножницы портновские, 10, 26 см, цвет чёрный"
                src="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476"
              />
            </div>
          </Image>
          <WithHint
            text="Добавить в избранное"
          >
            <ForwardRef
              checked={false}
              className="on-image-btn wish"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <ForwardRef(SvgHeart)
                className="button on-image-btn wish"
                data-testid="favorite-button"
                height={24}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="button"
                title="Добавить в избранное"
                width={24}
              >
                <svg
                  className="button on-image-btn wish"
                  data-testid="favorite-button"
                  height={24}
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="button"
                  title="Добавить в избранное"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.944 3.6h-.001c-2.7-1.879-5.978-1.165-7.943.793-1.964-1.958-5.241-2.683-7.943-.792C2.524 4.673 1.57 6.473 1.504 8.369c-.075 2.145.826 4.042 2.36 5.985 1.512 1.912 3.685 3.927 6.261 6.315l.086.08h.002l.098.091a2.466 2.466 0 003.371-.012l.107-.1.027-.025c2.602-2.407 4.795-4.436 6.316-6.361 1.534-1.941 2.435-3.838 2.364-5.982v-.001c-.065-1.886-1.02-3.686-2.552-4.758zM12 18.274c-2.278-2.105-4.14-3.84-5.45-5.401-1.358-1.62-2.047-2.968-2.047-4.261 0-1.774 1.285-3.07 2.999-3.07 1.055 0 2.103.553 2.72 1.39.38.516.996 1.017 1.78 1.017.789 0 1.406-.507 1.787-1.025.612-.831 1.657-1.382 2.71-1.382 1.713 0 2.998 1.296 2.998 3.07 0 1.293-.689 2.641-2.047 4.26-1.309 1.561-3.171 3.296-5.45 5.402z"
                  />
                </svg>
              </ForwardRef(SvgHeart)>
            </ForwardRef>
          </WithHint>
          <WithHint
            text="Быстрый просмотр"
          >
            <ForwardRef
              className="on-image-btn quick-view"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <ForwardRef(SvgQuickView)
                className="root on-image-btn quick-view"
                data-testid="quick-view-button"
                height={24}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="button"
                title="Быстрый просмотр"
                width={24}
              >
                <svg
                  className="root on-image-btn quick-view"
                  data-testid="quick-view-button"
                  height={24}
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="button"
                  title="Быстрый просмотр"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.94 22.06l-3.219-3.217-.339.22A8.956 8.956 0 0111.5 20.5a9 9 0 119-9c0 1.8-.528 3.476-1.438 4.882l-.22.34 3.219 3.217a1.5 1.5 0 01-2.122 2.122zM9 10h1V9a1.5 1.5 0 113 0v1h1a1.5 1.5 0 110 3h-1v1a1.5 1.5 0 11-3 0v-1H9a1.5 1.5 0 110-3zm2.5-4.5a6 6 0 100 12 6 6 0 000-12z"
                  />
                </svg>
              </ForwardRef(SvgQuickView)>
            </ForwardRef>
          </WithHint>
        </a>
        <div
          className="row prices-row"
        >
          <Price
            className="price"
            currencyGrapheme="₽"
            value={225}
          >
            <span
              className="root price"
            >
              225 ₽
            </span>
          </Price>
          <Price
            className="old-price"
            crossedOut={true}
            currencyGrapheme="₽"
            value={250}
          >
            <span
              className="root old-price crossed-out"
            >
              250 ₽
            </span>
          </Price>
        </div>
        <ForwardRef(Link)
          className="row link"
          color="gray87"
          href="https://www.sima-land.ru/123456"
        >
          <a
            className="link row link color__gray87 hover-color__gray54"
            href="https://www.sima-land.ru/123456"
          >
            Ножницы портновские, 10, 26 см, цвет чёрный
          </a>
        </ForwardRef(Link)>
      </ProductInfo>
      <div
        className="control-wrapper"
      >
        <InCartControl
          className="control"
          markupText="Комплектация + 50 ₽ при покупке до 20 шт"
          onAdd={[Function]}
          onChange={[Function]}
          onSubtract={[Function]}
        >
          <ForwardRef(Button)
            className="control control"
            onClick={[Function]}
            size="small"
          >
            <button
              className="control control button-base button-primary button-small text-button-small"
              onClick={[Function]}
            >
              В корзину
            </button>
          </ForwardRef(Button)>
        </InCartControl>
        <div
          className="bottom-text"
        >
          Комплектация + 50 ₽ при покупке до 20 шт
        </div>
      </div>
    </div>
  </ForwardRef>
</HoverCard>
`;
